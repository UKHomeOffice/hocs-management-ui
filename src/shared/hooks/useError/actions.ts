import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
