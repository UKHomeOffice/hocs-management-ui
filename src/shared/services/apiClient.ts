import axios from 'axios';

export default {
    createClient: () => {
        const client = axios.create();

        return {
            get: client.get,
            delete: client.delete,
            post: client.post,
            put: client.put
        };
    }
};
