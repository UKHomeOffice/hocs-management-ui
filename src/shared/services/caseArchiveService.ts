import axios from 'axios';
import ArchiveCaseModel from '../models/archiveCaseModel';

export const archiveCase = (archiveCaseModel: ArchiveCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/archive', archiveCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
