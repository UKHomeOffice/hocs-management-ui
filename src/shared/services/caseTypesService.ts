import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
