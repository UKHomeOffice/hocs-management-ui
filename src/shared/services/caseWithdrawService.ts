import axios from 'axios';
import WithdrawCaseModel from '../models/withdrawCaseModel';

export const withdrawCase = (withdrawCaseModel: WithdrawCaseModel) => new Promise((resolve, reject) => axios
    .post('/api/case/withdraw', withdrawCaseModel)
    .then(() => resolve(null))
    .catch(reason => reject(reason))
);
