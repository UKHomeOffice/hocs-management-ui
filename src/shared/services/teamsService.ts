import axios from 'axios';
import Item from '../models/item';
import Team from '../models/team';

export const getTeam = (teamId: string) => new Promise<Team>((resolve, reject) =>
    axios.get(`/api/teams/${teamId}`)
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeams = () => new Promise<[Item]>((resolve, reject) =>
    axios.get('/api/teams')
        .then(response => resolve(response.data))
        .catch(error => reject(error))
);

export const getTeamMembers = (teamId: string) => new Promise((resolve, reject) => axios
    .get(`/api/teams/${teamId}/members`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getTeamsForUser = (userId: string) => new Promise<[Item]>((resolve, reject) => axios
    .get(`/api/teams/${userId}/teams`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const createTeam = (team: Team) => new Promise((resolve, reject) => axios
    .post(`/api/teams/unit/${team.unitUUID}`, team)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
