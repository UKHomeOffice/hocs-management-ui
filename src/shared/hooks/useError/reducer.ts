import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
