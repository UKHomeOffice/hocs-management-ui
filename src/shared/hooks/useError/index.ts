import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
