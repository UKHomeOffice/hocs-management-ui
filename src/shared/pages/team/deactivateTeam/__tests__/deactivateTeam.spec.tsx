import { act, fireEvent, render, RenderResult, wait, waitForElement } from '@testing-library/react';

import { MemoryRouter } from 'react-router-dom';
import React from 'react';
import DeactivateTeam from '../deactivateTeam';
import { createBrowserHistory, History } from 'history';
import * as TeamsService from '../../../../services/teamsService';
import * as useError from '../../../../hooks/useError';


const mockSetTeamName = jest.fn();
const mockTeamName = '__TeamId__';
const getTeamSpy = jest.spyOn(TeamsService, 'getTeam').mockReturnValue(
    Promise.resolve({ displayName: '__displayName__', active: true, letterName: '', permissions: [], type: '' } )
);
const useStateSpy = jest.spyOn(React, 'useState').mockImplementation(() => [mockTeamName, mockSetTeamName]);
const updateTeamMock = jest.spyOn(TeamsService, 'updateTeam').mockImplementation(() => Promise.resolve(200));
const setErrorMessageSpy = jest.fn();
jest.spyOn(useError, 'default').mockImplementation(() => [{}, jest.fn(), jest.fn(), setErrorMessageSpy]);

let history: History<any>;

const match = {
    isExact: true,
    params: { teamId: '__teamId__' },
    path: '',
    url: ''
};

const location = {
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: {}
};

const renderComponent = async () => {
    history = createBrowserHistory();
    history.push = jest.fn();

    act(() => {
        result = render(
            <MemoryRouter>
                <DeactivateTeam history={history} match={match} location={location}/>
            </MemoryRouter>
        );
    });

};

let result: RenderResult;

beforeEach(async () => {
    renderComponent();
});

describe('when the deactivateTeam component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);

        expect(getTeamSpy).toHaveBeenCalledWith('__teamId__');
        expect(useStateSpy).toHaveBeenCalled();
        expect(result.container).toMatchSnapshot();
    });
});

describe('when the Deactivate Team button is clicked', () => {
    it('should send a request to deactivate the team and redirect to the team view', async () => {
        const deactivateButton = await waitForElement(async () => {
            return await result.findByText('Deactivate Team');
        });

        fireEvent.click(deactivateButton);

        expect(updateTeamMock).toHaveBeenCalledWith('__teamId__', { active: false });
        await wait(() => {
            expect(history.push).toHaveBeenCalledWith('/team-view/__teamId__', { successMessage: '__TeamId__ has been deactivated successfully' });
        });
    });

    it('should set an error when call to deactivate fails', async () => {
        updateTeamMock.mockImplementationOnce(() => Promise.reject({ response: { status: 500, data: { body: 'update failed' } } } ));

        const deactivateButton = await waitForElement(async () => {
            return await result.findByText('Deactivate Team');
        });

        fireEvent.click(deactivateButton);

        expect(updateTeamMock).toHaveBeenCalledWith('__teamId__', { active: false });
        await wait(() => {
            expect(setErrorMessageSpy).toHaveBeenCalled();
        });
    });
});