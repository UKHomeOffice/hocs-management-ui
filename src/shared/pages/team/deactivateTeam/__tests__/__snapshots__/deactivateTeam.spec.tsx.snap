// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the deactivateTeam component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/team-view/__teamId__"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Deactivating team
      </h1>
      <div
        class="govuk-body"
      >
        <p>
          You are deactivating the 
          <strong>
            __TeamId__
          </strong>
           team.
        </p>
        <p>
          New cases wonâ€™t be assigned to it, but existing cases are unaffected
        </p>
        <p>
          This means:
        </p>
        <ul>
          <li>
            cases already with this team will stay with the team
          </li>
          <li>
            existing cases with this team can still be worked
          </li>
          <li>
            this team's topics will remain active but without a recommended team
          </li>
          <li>
            the team will no longer appear in drop-downs
          </li>
          <li>
            caseworkers will be prevented from giving new cases to the team
          </li>
          <li>
            in-flight cases that passed through this team can still be returned to it
          </li>
          <li>
            caseworkers will remain part of the team
          </li>
        </ul>
        <p>
          In an emergency, cases might be routed to the DCU central team.
        </p>
        <p>
          You can reactivate at any time.
        </p>
      </div>
      <div
        class="govuk-form-group"
      >
        <button
          class="govuk-button govuk-!-margin-right-1"
          type="submit"
        >
          Deactivate Team
        </button>
        <a
          class="govuk-button govuk-button--secondary"
          href="/team-view/__teamId__"
        >
          Cancel
        </a>
      </div>
    </div>
  </div>
</div>
`;
